### PR TITLE
Move consensus and execution validation outside of onBlock

### PR DESCRIPTION
**Pull request by potuz**
This is a simpler alternative to #12561 

It moves consensus and execution validation outside of `onBlock` and acquires thus the forkchoice lock after validation. This allows for a simpler lock logic while still not keeping the lock during the bulk of the processing. 

This method also cleans up the path to parallelizing consensus and execution validation 

Locking during init sync stays the same, the node is essentially locked to do anything else while processing batches.

### CODE REVIEW

File: beacon-chain/blockchain/execution_engine.go [hunk 1]
```diff
@@ -182,21 +182,21 @@ func (s *Service) getPayloadHash(ctx context.Context, root []byte) ([32]byte, er
 
 // notifyNewPayload signals execution engine on a new payload.
 // It returns true if the EL has returned VALID for the block
-func (s *Service) notifyNewPayload(ctx context.Context, postStateVersion int,
-	postStateHeader interfaces.ExecutionData, blk interfaces.ReadOnlySignedBeaconBlock) (bool, error) {
+func (s *Service) notifyNewPayload(ctx context.Context, preStateVersion int,
```
Comment (potuz) [inline beacon-chain/blockchain/execution_engine.go:185]: This is required so that execution validation is not dependent on consensus validation

File: beacon-chain/blockchain/execution_engine.go [hunk 2]
```diff
@@ -182,21 +182,21 @@ func (s *Service) getPayloadHash(ctx context.Context, root []byte) ([32]byte, er
...
-	if blocks.IsPreBellatrixVersion(postStateVersion) {
+	if blocks.IsPreBellatrixVersion(preStateVersion) {
 		return true, nil
 	}
 	if err := consensusblocks.BeaconBlockIsNil(blk); err != nil {
 		return false, err
 	}
 	body := blk.Block().Body()
-	enabled, err := blocks.IsExecutionEnabledUsingHeader(postStateHeader, body)
+	enabled, err := blocks.IsExecutionEnabledUsingHeader(preStateHeader, body)
```
Comment (potuz) [inline beacon-chain/blockchain/execution_engine.go:199]: This is the only place where there is a logic change: it may be that the execution payload right after the block transition is enabled but not before. But this is checked in `validateMergeBlock` anyway.

File: beacon-chain/blockchain/process_block_helpers.go [hunk 3]
```diff
@@ -209,35 +209,35 @@ func (s *Service) fillInForkChoiceMissingBlocks(ctx context.Context, blk interfa
...
+// called in the background
+func (s *Service) insertFinalizedDeposits(fRoot [32]byte) {
+	ctx, span := trace.StartSpan(context.Background(), "blockChain.insertFinalizedDeposits")
 	defer span.End()
 
 	// Update deposit cache.
 	finalizedState, err := s.cfg.StateGen.StateByRoot(ctx, fRoot)
 	if err != nil {
-		return errors.Wrap(err, "could not fetch finalized state")
+		log.WithError(err).Error("could not fetch finalized state")
```
Comment (potuz) [inline beacon-chain/blockchain/process_block_helpers.go:221]: these errors should not prevent block insertion

File: beacon-chain/blockchain/process_block_test.go [hunk 4]
```diff
@@ -41,103 +41,6 @@ import (
 	logTest "github.com/sirupsen/logrus/hooks/test"
 )
 
-func TestStore_OnBlock(t *testing.T) {
```
Comment (potuz) [inline beacon-chain/blockchain/process_block_test.go:44]: This test does no longer test `onBlock` but rather the previous helpers that are tested elsewhere

File: beacon-chain/blockchain/receive_block.go [hunk 5]
```diff
@@ -226,3 +270,114 @@ func (s *Service) checkSaveHotStateDB(ctx context.Context) error {
...
+// This performs the state transition function and returns the poststate or an
+// error if the block fails to verify the consensus rules
+func (s *Service) validateStateTransition(ctx context.Context, preState state.BeaconState, signed interfaces.ReadOnlySignedBeaconBlock) (state.BeaconState, error) {
+	b := signed.Block()
+	// Verify that the parent block is in forkchoice
+	parentRoot := b.ParentRoot()
+	if !s.InForkchoice(parentRoot) {
+		return nil, ErrNotDescendantOfFinalized
+	}
+	stateTransitionStartTime := time.Now()
```
Comment (potuz) [inline beacon-chain/blockchain/receive_block.go:283]: Notice that the total time in chain log will be changed now.

File: beacon-chain/blockchain/process_block.go [hunk 6]
```diff
@@ -92,7 +91,7 @@ var initialSyncBlockCacheSize = uint64(2 * params.BeaconConfig().SlotsPerEpoch)
 //	         ancestor_at_finalized_slot = get_ancestor(store, store.justified_checkpoint.root, finalized_slot)
 //	         if ancestor_at_finalized_slot != store.finalized_checkpoint.root:
 //	             store.justified_checkpoint = state.current_justified_checkpoint
-func (s *Service) onBlock(ctx context.Context, signed interfaces.ReadOnlySignedBeaconBlock, blockRoot [32]byte) error {
+func (s *Service) onBlock(ctx context.Context, signed interfaces.ReadOnlySignedBeaconBlock, blockRoot [32]byte, postState state.BeaconState, isValidPayload bool) error {
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go:94]: I really dont like we have to pass `isValidPayload`. I think we can do better here but let's do it in another PR

Comment (terencechain) [inline beacon-chain/blockchain/process_block.go:94]: Can you update `onBlock` comments to make it more reflective?  or suggestion to rename `onBlock` to something else

File: beacon-chain/blockchain/execution_engine.go [hunk 7]
```diff
@@ -233,6 +233,8 @@ func (s *Service) notifyNewPayload(ctx context.Context, postStateVersion int,
 // reportInvalidBlock deals with the event that an invalid block was detected by the execution layer
 func (s *Service) reportInvalidBlock(ctx context.Context, root, parentRoot, lvh [32]byte) error {
 	newPayloadInvalidNodeCount.Inc()
+	s.cfg.ForkChoiceStore.Lock()
+	defer s.cfg.ForkChoiceStore.Unlock()
```
Comment (terencechain) [inline beacon-chain/blockchain/execution_engine.go:237]: I am afraid this can be missused, I can see someone adding this `reportInvalidBlock` somewhere where it's already locked. What do you think about just removing `reportInvalidBlock` and have it inside `validateExecutionOnBlock`. I dont think this needs to be another function

**Review by terencechain — CHANGES_REQUESTED** [review]

File: beacon-chain/blockchain/process_block.go (hunk 6 above)
Comment (potuz) [reply beacon-chain/blockchain/process_block.go:94]: I'm planning on breaking further `onBlock` into smaller helpers, but that requires first breaking FCU which will take some time

File: beacon-chain/blockchain/execution_engine.go (hunk 7 above)
Comment (potuz) [reply beacon-chain/blockchain/execution_engine.go]: kept a different function but addded a chain-info helper that locks.

File: beacon-chain/blockchain/execution_engine.go [hunk 8]
```diff
@@ -233,7 +233,7 @@ func (s *Service) notifyNewPayload(ctx context.Context, postStateVersion int,
 // reportInvalidBlock deals with the event that an invalid block was detected by the execution layer
 func (s *Service) reportInvalidBlock(ctx context.Context, root, parentRoot, lvh [32]byte) error {
```
Comment (terencechain) [inline beacon-chain/blockchain/execution_engine.go:234]: I know this is an old name, but I think we should change it to `pruneInvalidBlock`

File: beacon-chain/blockchain/receive_block.go [hunk 9]
```diff
@@ -226,3 +277,110 @@ func (s *Service) checkSaveHotStateDB(ctx context.Context) error {
...
+	isValidPayload, err := s.notifyNewPayload(ctx, ver, header, signed)
+	if err != nil {
+		if IsInvalidBlock(err) && InvalidBlockLVH(err) != [32]byte{} {
+			return false, s.reportInvalidBlock(ctx, blockRoot, signed.Block().ParentRoot(), InvalidBlockLVH(err))
+		}
+		return false, errors.Wrap(err, "could not validate new payload")
+	}
+	if signed.Version() < version.Capella && isValidPayload {
+		if err := s.validateMergeTransitionBlock(ctx, ver, header, signed); err != nil {
+			return isValidPayload, err
```
Comment (terencechain) [inline beacon-chain/blockchain/receive_block.go:382]: Shouldn't `validateMergeTransitionBlock` error be wrapped with `invalidBlock{}` as well?

File: beacon-chain/blockchain/process_block.go [hunk 10]
```diff
@@ -101,51 +100,6 @@ func (s *Service) onBlock(ctx context.Context, signed interfaces.ReadOnlySignedB
...
-	}
-	if signed.Version() < version.Capella && isValidPayload {
-		if err := s.validateMergeTransitionBlock(ctx, preStateVersion, preStateHeader, signed); err != nil {
-			return err
-		}
-	}
-
 	if err := s.savePostStateInfo(ctx, blockRoot, signed, postState); err != nil {
 		return err
 	}
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go:105]: Why not move `savePostStateInfo` outside as well. This is not a cheap operation

Comment (potuz) [reply beacon-chain/blockchain/process_block.go:105]: Because I wouldn't gain anything: if I release the lock before it then it starts to lol like the previous PR. The reason I needed to move finalization and justification out is to not have to pass two states or checkpoints around

Comment (terencechain) [reply beacon-chain/blockchain/process_block.go:105]: `savePostStateInfo` doesn't require the lock right, so everything can be done outside. That said, this is only 50-100ms so I'm ok not doing it

File: beacon-chain/blockchain/receive_block.go (hunk 9 above)
Comment (terencechain) [reply beacon-chain/blockchain/receive_block.go:382]: Ignore. It is wrapped in `validateMergeBlock`

File: beacon-chain/blockchain/execution_engine.go (hunk 1 above)
Comment (nisdas) [reply beacon-chain/blockchain/execution_engine.go:185]: This changes it significantly no ? from pre-state checks to now post-state

File: beacon-chain/blockchain/process_block_helpers.go (hunk 3 above)
Comment (nisdas) [reply beacon-chain/blockchain/process_block_helpers.go:221]: They don't there are in an alternate routine for that purpose

Comment (rkapka) [reply beacon-chain/blockchain/process_block_helpers.go:221]: If we end up changing errors to logs, we should capitalize messages

File: beacon-chain/blockchain/receive_block.go [hunk 11]
```diff
@@ -226,3 +277,110 @@ func (s *Service) checkSaveHotStateDB(ctx context.Context) error {
...
+		indexedAtt, err := attestation.ConvertToIndexed(ctx, att, committee)
+		if err != nil {
+			log.WithError(err).Error("Could not convert to indexed attestation")
+			return
+		}
+		s.cfg.SlasherAttestationsFeed.Send(indexedAtt)
+	}
+}
+
+// validateExecutionOnBlock notifies the engine of the incoming block execution payload returns true if the payload is valid
```
Comment (rkapka) [inline beacon-chain/blockchain/receive_block.go:371]:
```suggestion
// validateExecutionOnBlock notifies the engine of the incoming block execution payload and returns true if the payload is valid
```

File: beacon-chain/blockchain/receive_block.go [hunk 12]
```diff
@@ -226,3 +277,110 @@ func (s *Service) checkSaveHotStateDB(ctx context.Context) error {
...
+	stateTransitionStartTime := time.Now()
+	postState, err := transition.ExecuteStateTransition(ctx, preState, signed)
+	if err != nil {
+		return nil, invalidBlock{error: err}
+	}
+	stateTransitionProcessingTime.Observe(float64(time.Since(stateTransitionStartTime).Milliseconds()))
+	return postState, nil
+}
+
+func (s *Service) updateJustificationOnBlock(ctx context.Context, preState, postState state.BeaconState, preJustifiedEpoch primitives.Epoch) error {
```
Comment (rkapka) [inline beacon-chain/blockchain/receive_block.go:299]: You have a godoc for `updateFinalizationOnBlock`, so it would be nice to add one here for symmetry

File: beacon-chain/blockchain/receive_block.go [hunk 13]
```diff
@@ -226,3 +277,110 @@ func (s *Service) checkSaveHotStateDB(ctx context.Context) error {
 
 	return s.cfg.StateGen.DisableSaveHotStateToDB(ctx)
 }
+
+// This performs the state transition function and returns the poststate or an
+// error if the block fails to verify the consensus rules
+func (s *Service) validateStateTransition(ctx context.Context, preState state.BeaconState, signed interfaces.ReadOnlySignedBeaconBlock) (state.BeaconState, error) {
```
Comment (rkapka) [inline beacon-chain/blockchain/receive_block.go:283]: I would rename this function to `transitionState` because that's the main point, not the validation

File: beacon-chain/blockchain/process_block_helpers.go (hunk 3 above)
Comment (potuz) [reply beacon-chain/blockchain/process_block_helpers.go:221]: This is a log with an error, I treat it thus the same as an error message.

Comment (nisdas) [reply beacon-chain/blockchain/process_block_helpers.go:221]: But its not the same, the finalized state will be nil if there is an error and we will panic due to accessing a nil pointer.

File: beacon-chain/blockchain/receive_block.go [hunk 14]
```diff
@@ -226,3 +277,110 @@ func (s *Service) checkSaveHotStateDB(ctx context.Context) error {
...
+	return nil
+}
+
+// updateFinalizationOnBlock performs some duties when the incoming block
+// changes the finalized checkpoint. It returns true when this has happened.
+func (s *Service) updateFinalizationOnBlock(ctx context.Context, preState, postState state.BeaconState, preFinalizedEpoch primitives.Epoch) (bool, error) {
+	preStateFinalizedEpoch := preState.FinalizedCheckpoint().Epoch
+	postStateFinalizedEpoch := postState.FinalizedCheckpoint().Epoch
+	finalized := s.cfg.ForkChoiceStore.FinalizedCheckpoint()
+	if finalized.Epoch > preFinalizedEpoch || (finalized.Epoch == postStateFinalizedEpoch && finalized.Epoch > preStateFinalizedEpoch) {
```
Comment (nisdas) [inline beacon-chain/blockchain/receive_block.go:319]: Is this the same ? 
```
// Save finalized check point to db and more.
	postStateFinalizedEpoch := postState.FinalizedCheckpoint().Epoch
	finalized := s.cfg.ForkChoiceStore.FinalizedCheckpoint()
	if finalized.Epoch > currStoreFinalizedEpoch || (finalized.Epoch == postStateFinalizedEpoch && finalized.Epoch > preStateFinalizedEpoch)
```

Previously we were comparing with the store's finalized epoch, while now its with the prestate

File: beacon-chain/blockchain/receive_block.go [hunk 15]
```diff
@@ -226,3 +277,110 @@ func (s *Service) checkSaveHotStateDB(ctx context.Context) error {
...
+	}
+	stateTransitionProcessingTime.Observe(float64(time.Since(stateTransitionStartTime).Milliseconds()))
+	return postState, nil
+}
+
+func (s *Service) updateJustificationOnBlock(ctx context.Context, preState, postState state.BeaconState, preJustifiedEpoch primitives.Epoch) error {
+	justified := s.cfg.ForkChoiceStore.JustifiedCheckpoint()
+	preStateJustifiedEpoch := preState.CurrentJustifiedCheckpoint().Epoch
+	postStateJustifiedEpoch := postState.CurrentJustifiedCheckpoint().Epoch
+	if justified.Epoch > preJustifiedEpoch || (justified.Epoch == postStateJustifiedEpoch && justified.Epoch > preStateJustifiedEpoch) {
```
Comment (nisdas) [inline beacon-chain/blockchain/receive_block.go:303]:
```
postStateJustifiedEpoch := postState.CurrentJustifiedCheckpoint().Epoch
	if justified.Epoch > currStoreJustifiedEpoch || (justified.Epoch == postStateJustifiedEpoch && justified.Epoch > preStateJustifiedEpoch)
```

Same here for justification, are we meant to use prestate here ?

File: beacon-chain/blockchain/receive_block.go (hunk 14 above)
Comment (potuz) [reply beacon-chain/blockchain/receive_block.go:319]: yeah we pass these as 
```
	// Save current justified and finalized epochs for future use.
	currStoreJustifiedEpoch := s.CurrentJustifiedCheckpt().Epoch
	currStoreFinalizedEpoch := s.FinalizedCheckpt().Epoch
```
In lines 62-64 of `receive_block.go`

**Review by nisdas — APPROVED** [review]

**Review by terencechain — DISMISSED** [review]

**Review by terencechain — APPROVED** [review]

**Review by nisdas — APPROVED** [review]

**Review by terencechain — APPROVED** [review]
